# Leopard challenge Regression

## Загрузка данных
Данные беру из leopard-challenge-classification.zip
https://drive.google.com/file/d/1OYjGYeYQzKI3mE02ymtkrbRJb5EurpL7/view?usp=sharing

# Colab

In [1]:
! gdown --id 1OYjGYeYQzKI3mE02ymtkrbRJb5EurpL7

"gdown" ­Ґ пў«пҐвбп ў­гв७­Ґ© Ё«Ё ў­Ґи­Ґ©
Є®¬ ­¤®©, ЁбЇ®«­пҐ¬®© Їа®Ја ¬¬®© Ё«Ё Ї ЄҐв­л¬ д ©«®¬.


! unzip /content/leopard-challenge-classification.zip -d /content/leopard-challenge-classification/

In [9]:
! unzip / content / leopard-challenge-classification.zip -d / content / leopard_challenge_classification /

"unzip" ­Ґ пў«пҐвбп ў­гв७­Ґ© Ё«Ё ў­Ґи­Ґ©
Є®¬ ­¤®©, ЁбЇ®«­пҐ¬®© Їа®Ја ¬¬®© Ё«Ё Ї ЄҐв­л¬ д ©«®¬.


In [32]:
import pandas as pd

In [1]:
train = pd.read_csv('/content/leopard_challenge_regression/train.csv')

NameError: name 'pd' is not defined

In [1]:
train.head()

NameError: name 'train' is not defined

In [ ]:
!pip install phik

!pip install feature_engine

!pip install catboost

# Pycharm

In [1]:
from zipfile import ZipFile

with ZipFile('leopard-challenge-classification.zip', 'r') as zf:
    zf.extractall()

# 1. Задача

Дан набор данных с основными клиническими показателями людей.
Каждая строчка в датасете - клинические показатели одного человека - рост, вес, зрение и т.д.
**Задача:** Определить, курит человек или нет, на основании его показателей.

# 2. Импорт библиотек

In [3]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import missingno as msno
import phik
import folium
from feature_engine.selection import SmartCorrelatedSelection
from feature_engine.imputation import RandomSampleImputer
from sklearn.metrics import mean_absolute_percentage_error, mean_absolute_error, log_loss
from sklearn.model_selection import train_test_split
from catboost import CatBoostClassifier, Pool, cv
import joblib
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier
import warnings
warnings.filterwarnings("ignore")

# plt.style.use('dark_background')
plt.style.use('seaborn-colorblind')
pd.set_option('display.max_columns', 100)

# 3. Загрузка Датасета

In [33]:
df = pd.read_csv('train.csv')

In [34]:
df.head()

ID  age  height(cm)  weight(kg)  waist(cm)  eyesight(left)  \
0   2791   65         170          75       91.0             0.6   
1  18280   35         170          85       97.0             1.5   
2   1618   70         165          55       75.0             0.8   
3  16856   35         180          85       83.0             1.2   
4   5885   25         170          65       80.0             1.5   

   eyesight(right)  hearing(left)  hearing(right)  systolic  relaxation  \
0              0.9            1.0             1.0     122.0        79.0   
1              1.5            1.0             1.0     138.0        88.0   
2              1.0            1.0             1.0     115.0        63.0   
3              1.0            1.0             1.0     130.0        80.0   
4              1.2            1.0             1.0     135.0        75.0   

   fasting blood sugar  Cholesterol  triglyceride   HDL    LDL  hemoglobin  \
0                107.0        119.0          51.0  71.0   52.0        14.0   
1                117.0        204.0         239.0  46.0  120.0        14.6   
2                128.0        165.0          60.0  68.0   85.0        14.7   
3                100.0        209.0         155.0  53.0  125.0        17.2   
4                 94.0        153.0          73.0  58.0   80.0        16.5   

   Urine protein  serum creatinine   AST   ALT    Gtp oral  dental caries  \
0            3.0               1.2  20.0  16.0  122.0    Y              0   
1            1.0               0.7  20.0  25.0   40.0    Y              0   
2            1.0               0.7  15.0  13.0   19.0    Y              0   
3            1.0               0.8  22.0  18.0   17.0    Y              0   
4            1.0               1.0  20.0  13.0   14.0    Y              0   

  tartar  smoking  
0      N        0  
1      Y        1  
2      Y        0  
3      Y        0  
4      Y        1

In [35]:
df.shape

(13863, 26)

In [36]:
df_test = pd.read_csv('test.csv')

In [37]:
df.head()

ID  age  height(cm)  weight(kg)  waist(cm)  eyesight(left)  \
0   2791   65         170          75       91.0             0.6   
1  18280   35         170          85       97.0             1.5   
2   1618   70         165          55       75.0             0.8   
3  16856   35         180          85       83.0             1.2   
4   5885   25         170          65       80.0             1.5   

   eyesight(right)  hearing(left)  hearing(right)  systolic  relaxation  \
0              0.9            1.0             1.0     122.0        79.0   
1              1.5            1.0             1.0     138.0        88.0   
2              1.0            1.0             1.0     115.0        63.0   
3              1.0            1.0             1.0     130.0        80.0   
4              1.2            1.0             1.0     135.0        75.0   

   fasting blood sugar  Cholesterol  triglyceride   HDL    LDL  hemoglobin  \
0                107.0        119.0          51.0  71.0   52.0        14.0   
1                117.0        204.0         239.0  46.0  120.0        14.6   
2                128.0        165.0          60.0  68.0   85.0        14.7   
3                100.0        209.0         155.0  53.0  125.0        17.2   
4                 94.0        153.0          73.0  58.0   80.0        16.5   

   Urine protein  serum creatinine   AST   ALT    Gtp oral  dental caries  \
0            3.0               1.2  20.0  16.0  122.0    Y              0   
1            1.0               0.7  20.0  25.0   40.0    Y              0   
2            1.0               0.7  15.0  13.0   19.0    Y              0   
3            1.0               0.8  22.0  18.0   17.0    Y              0   
4            1.0               1.0  20.0  13.0   14.0    Y              0   

  tartar  smoking  
0      N        0  
1      Y        1  
2      Y        0  
3      Y        0  
4      Y        1

In [38]:
df_test.shape

(5942, 25)

# 4. Получение информации о Датасете

## 4.1 Подготовка таблиц

### Перевод названий колонок в строчные буквы и замена пробелов

In [39]:
df

ID  age  height(cm)  weight(kg)  waist(cm)  eyesight(left)  \
0       2791   65         170          75       91.0             0.6   
1      18280   35         170          85       97.0             1.5   
2       1618   70         165          55       75.0             0.8   
3      16856   35         180          85       83.0             1.2   
4       5885   25         170          65       80.0             1.5   
...      ...  ...         ...         ...        ...             ...   
13858   7835   45         175          85       94.0             0.8   
13859   9335   40         170          75       86.0             1.2   
13860  18936   65         170          70       85.0             0.6   
13861  11064   30         160          80       89.0             1.5   
13862   2646   60         170          70       86.0             1.5   

       eyesight(right)  hearing(left)  hearing(right)  systolic  relaxation  \
0                  0.9            1.0             1.0     122.0        79.0   
1                  1.5            1.0             1.0     138.0        88.0   
2                  1.0            1.0             1.0     115.0        63.0   
3                  1.0            1.0             1.0     130.0        80.0   
4                  1.2            1.0             1.0     135.0        75.0   
...                ...            ...             ...       ...         ...   
13858              0.8            1.0             1.0     127.0        71.0   
13859              1.0            1.0             1.0     134.0        88.0   
13860              0.7            1.0             1.0     131.0        82.0   
13861              1.5            1.0             1.0     120.0        80.0   
13862              1.5            1.0             1.0     135.0        85.0   

       fasting blood sugar  Cholesterol  triglyceride   HDL    LDL  \
0                    107.0        119.0          51.0  71.0   52.0   
1                    117.0        204.0         239.0  46.0  120.0   
2                    128.0        165.0          60.0  68.0   85.0   
3                    100.0        209.0         155.0  53.0  125.0   
4                     94.0        153.0          73.0  58.0   80.0   
...                    ...          ...           ...   ...    ...   
13858                 99.0        249.0         179.0  49.0  164.0   
13859                 97.0        247.0         183.0  62.0  148.0   
13860                 99.0        180.0          88.0  47.0  115.0   
13861                 92.0        172.0         136.0  49.0   96.0   
13862                 90.0        159.0          70.0  55.0   90.0   

       hemoglobin  Urine protein  serum creatinine   AST   ALT    Gtp oral  \
0            14.0            3.0               1.2  20.0  16.0  122.0    Y   
1            14.6            1.0               0.7  20.0  25.0   40.0    Y   
2            14.7            1.0               0.7  15.0  13.0   19.0    Y   
3            17.2            1.0               0.8  22.0  18.0   17.0    Y   
4            16.5            1.0               1.0  20.0  13.0   14.0    Y   
...           ...            ...               ...   ...   ...    ...  ...   
13858        16.4            1.0               0.9  17.0  17.0   19.0    Y   
13859        16.7            1.0               0.9  15.0  12.0   36.0    Y   
13860        15.7            1.0               0.8  12.0   8.0   24.0    Y   
13861        14.8            1.0               0.8  26.0  35.0   31.0    Y   
13862        14.9            1.0               1.0  56.0  85.0   32.0    Y   

       dental caries tartar  smoking  
0                  0      N        0  
1                  0      Y        1  
2                  0      Y        0  
3                  0      Y        0  
4                  0      Y        1  
...              ...    ...      ...  
13858              0      N        0  
13859              0      Y        1  
13860              1      Y        1  
13861              0      N 

In [40]:
df.columns = df.columns.str.replace(' ', '_').str.lower()
df_test.columns = df_test.columns.str.replace(' ', '_').str.lower()

In [42]:
df_test

id  age  height(cm)  weight(kg)  waist(cm)  eyesight(left)  \
0     19200   40         170          75       87.0             1.2   
1      7784   25         170          65       78.4             0.9   
2     11682   50         180         105      109.3             0.9   
3      9949   60         160          60       81.0             0.5   
4      9378   45         175          70       83.0             0.9   
...     ...  ...         ...         ...        ...             ...   
5937  17154   40         170          75       85.5             1.0   
5938  19366   25         165          60       73.5             0.9   
5939   1170   40         180          80       94.0             1.2   
5940   9358   20         170          70       84.5             1.2   
5941  14302   45         165          45       71.0             0.9   

      eyesight(right)  hearing(left)  hearing(right)  systolic  relaxation  \
0                 1.0            1.0             1.0     140.0        90.0   
1                 0.9            1.0             1.0     118.0        78.0   
2                 1.0            1.0             1.0     130.0        90.0   
3                 0.5            1.0             1.0     118.0        76.0   
4                 1.0            1.0             1.0     130.0        80.0   
...               ...            ...             ...       ...         ...   
5937              0.1            1.0             1.0     113.0        65.0   
5938              1.0            1.0             1.0     115.0        70.0   
5939              1.0            1.0             1.0     126.0        69.0   
5940              1.5            1.0             1.0     110.0        70.0   
5941              0.9            1.0             2.0     124.0        80.0   

      fasting_blood_sugar  cholesterol  triglyceride    hdl    ldl  \
0                   131.0        225.0         216.0   52.0  129.0   
1                   101.0        170.0          33.0   72.0   91.0   
2                    96.0        162.0         144.0   71.0   62.0   
3                    97.0        172.0          91.0   46.0  108.0   
4                   100.0        191.0         365.0   34.0   84.0   
...                   ...          ...           ...    ...    ...   
5937                108.0        153.0          87.0   48.0   88.0   
5938                 79.0        235.0          69.0   50.0  171.0   
5939                108.0        170.0          65.0   52.0  105.0   
5940                 92.0        156.0          79.0   48.0   92.0   
5941                 85.0        207.0          71.0  101.0   92.0   

      hemoglobin  urine_protein  serum_creatinine   ast   alt   gtp oral  \
0           16.4            1.0               1.1  25.0  58.0  43.0    Y   
1           11.9            1.0               0.9  33.0  66.0  22.0    Y   
2           15.4            1.0               0.9  19.0  24.0  32.0    Y   
3           16.2            1.0               1.2  34.0  37.0  26.0    Y   
4           15.2            1.0               0.7  27.0  25.0  26.0    Y   
...          ...            ...               ...   ...   ...   ...  ...   
5937        14.2            1.0               1.1  14.0  18.0  15.0    Y   
5938        14.4            1.0               0.9  20.0  12.0  17.0    Y   
5939        14.5            1.0               0.8  28.0  33.0  25.0    Y   
5940        16.1            1.0               0.9  28.0  21.0  21.0    Y   
5941        15.2            1.0               0.9  22.0  14.0  20.0    Y   

      dental_caries tartar  
0                 0      Y  
1                 0      N  
2                 0      Y  
3                 0      Y  
4                 0      N  
...             ...    ...  
5937              1      N  
5938              1      Y  
5939              0      N  
5940              1      Y  
5941              0      N  

[5942 rows x 25 columns]

## 4.2 Основная информация о датасете

In [44]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13863 entries, 0 to 13862
Data columns (total 26 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   13863 non-null  int64  
 1   age                  13863 non-null  int64  
 2   height(cm)           13863 non-null  int64  
 3   weight(kg)           13863 non-null  int64  
 4   waist(cm)            13863 non-null  float64
 5   eyesight(left)       13863 non-null  float64
 6   eyesight(right)      13863 non-null  float64
 7   hearing(left)        13863 non-null  float64
 8   hearing(right)       13863 non-null  float64
 9   systolic             13863 non-null  float64
 10  relaxation           13863 non-null  float64
 11  fasting_blood_sugar  13863 non-null  float64
 12  cholesterol          13863 non-null  float64
 13  triglyceride         13863 non-null  float64
 14  hdl                  13863 non-null  float64
 15  ldl                  13863 non-null 

**Обучающая таблица состоит из:** 13863 строк и 26 столбцов

In [45]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5942 entries, 0 to 5941
Data columns (total 25 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   5942 non-null   int64  
 1   age                  5942 non-null   int64  
 2   height(cm)           5942 non-null   int64  
 3   weight(kg)           5942 non-null   int64  
 4   waist(cm)            5942 non-null   float64
 5   eyesight(left)       5942 non-null   float64
 6   eyesight(right)      5942 non-null   float64
 7   hearing(left)        5942 non-null   float64
 8   hearing(right)       5942 non-null   float64
 9   systolic             5942 non-null   float64
 10  relaxation           5942 non-null   float64
 11  fasting_blood_sugar  5942 non-null   float64
 12  cholesterol          5942 non-null   float64
 13  triglyceride         5942 non-null   float64
 14  hdl                  5942 non-null   float64
 15  ldl                  5942 non-null   f

**Тестовая таблица состоит из:** 5942 строк и 25 столбцов

## 4.3 Пропущенные значения

In [46]:
df.isna().mean().sort_values(ascending=False)

id                     0.0
age                    0.0
tartar                 0.0
dental_caries          0.0
oral                   0.0
gtp                    0.0
alt                    0.0
ast                    0.0
serum_creatinine       0.0
urine_protein          0.0
hemoglobin             0.0
ldl                    0.0
hdl                    0.0
triglyceride           0.0
cholesterol            0.0
fasting_blood_sugar    0.0
relaxation             0.0
systolic               0.0
hearing(right)         0.0
hearing(left)          0.0
eyesight(right)        0.0
eyesight(left)         0.0
waist(cm)              0.0
weight(kg)             0.0
height(cm)             0.0
smoking                0.0
dtype: float64

In [47]:
df_test.isna().mean().sort_values(ascending=False)

id                     0.0
triglyceride           0.0
dental_caries          0.0
oral                   0.0
gtp                    0.0
alt                    0.0
ast                    0.0
serum_creatinine       0.0
urine_protein          0.0
hemoglobin             0.0
ldl                    0.0
hdl                    0.0
cholesterol            0.0
age                    0.0
fasting_blood_sugar    0.0
relaxation             0.0
systolic               0.0
hearing(right)         0.0
hearing(left)          0.0
eyesight(right)        0.0
eyesight(left)         0.0
waist(cm)              0.0
weight(kg)             0.0
height(cm)             0.0
tartar                 0.0
dtype: float64

**Пропуски в данных отсутствуют!!**

## 4.4 Дубликаты

In [48]:
df.duplicated().sum()

0

In [49]:
df_test.duplicated().sum()

0

**Дубликаты отсутствуют!**

## 4.5 Числовые переменные

In [50]:
df.describe()

id           age    height(cm)    weight(kg)     waist(cm)  \
count  13863.000000  13863.000000  13863.000000  13863.000000  13863.000000   
mean    9932.891654     41.743851    169.618409     71.270288     84.870627   
std     5715.739532     13.624383      6.493959     11.195751      8.194957   
min        1.000000     20.000000    145.000000     40.000000     51.000000   
25%     4985.500000     30.000000    165.000000     65.000000     79.200000   
50%     9956.000000     40.000000    170.000000     70.000000     84.400000   
75%    14865.500000     50.000000    175.000000     80.000000     90.000000   
max    19804.000000     85.000000    190.000000    130.000000    128.000000   

       eyesight(left)  eyesight(right)  hearing(left)  hearing(right)  \
count    13863.000000     13863.000000   13863.000000    13863.000000   
mean         1.057304         1.053740       1.028060        1.027411   
std          0.514618         0.514613       0.165151        0.163284   
min          0.100000         0.100000       1.000000        1.000000   
25%          0.800000         0.800000       1.000000        1.000000   
50%          1.000000         1.000000       1.000000        1.000000   
75%          1.200000         1.200000       1.000000        1.000000   
max          9.900000         9.900000       2.000000        2.000000   

           systolic    relaxation  fasting_blood_sugar   cholesterol  \
count  13863.000000  13863.000000         13863.000000  13863.000000   
mean     123.430643     77.166703            99.937387    193.864387   
std       12.964684      9.459288            20.682085     35.658783   
min       74.000000     40.000000            48.000000     84.000000   
25%      115.000000     70.000000            90.000000    169.000000   
50%      122.000000     78.000000            96.000000    192.000000   
75%      131.000000     83.000000           104.000000    216.000000   
max      240.000000    140.000000           475.000000    441.000000   

       triglyceride           hdl           ldl    hemoglobin  urine_protein  \
count  13863.000000  13863.000000  13863.000000  13863.000000   13863.000000   
mean     130.917767     54.204357    114.435476     15.402200       1.093775   
std       71.348298     13.168432     42.050428      1.052598       0.421272   
min       16.000000      4.000000      7.000000      6.300000       1.000000   
25%       78.000000     45.000000     91.000000     14.800000       1.000000   
50%      113.000000     53.000000    113.000000     15.400000       1.000000   
75%      166.000000     61.000000    135.000000     16.100000       1.000000   
max      466.000000    359.000000   1340.000000     19.600000       6.000000   

       serum_creatinine           ast           alt           gtp  \
count      13863.000000  13863.000000  13863.000000  13863.000000   
mean           0.981808     27.597562     30.772921     43.442978   
std            0.240337     21.559820     36.441688     50.342097   
min            0.100000      7.000000      2.000000      2.000000   
25%            0.900000     20.000000     18.000000     20.000000   
50%            1.000000     24.000000     24.000000     29.000000   
75%            1.100000     30.000000     35.000000     48.000000   
max           10.000000    981.000000   2914.000000    999.000000   

       dental_caries       smoking  
count   13863.000000  13863.000000  
mean        0.212075      0.201976  
std         0.408793      0.401489  
min         0.000000      0.000000  
25%         0.000000      0.000000  
50%         0.000000      0.000000  
75%         0.000000      0.000000  
max         1.000000      1.000000

#### Описание числовых переменных
**1. Возраст** Минимальный - 20, Максимальный - 85, Средний ≈41, Медианный ≈40
**2. Рост (см)** Минимальный - 145, Максимальны 190, Средний ≈169, Медианный ≈170
**3. Вес (кг)** Минимальный - 40, Максимальный 130, Средний ≈71.2, Медианный ≈70
**4. Окружность талии (см)** Минимальная - 51, Максимальная - 128, Средняя ≈84.8 Медианная ≈84.4
**5. Зрение (левый глаз)** Минимальное 0.1, Максимальное 9.9 (**ВЫБРОС?**) Среднее ≈1.05, Медианное ≈1
**6. Зрение (правый глаз)** Минимальное 0.1, Максимальное 9.9 (**ВЫБРОС?**) Среднее ≈1.05, Медианное ≈1
**7. Слух (левое ухо)** Минимальный 0.16, Максимальный 2, Средний ≈1.02, Медианный ≈1
**8. Слух (правое ухо)** Минимальный 0.16, Максимальный 2, Средний ≈1.02, Медианный ≈1
**9. Систолическое давление** Минимальное 12.9(**ВЫБРОС?**), Максимальное 240(**ВЫБРОС?**) Среднее ≈123.4, Медианное ≈122
**9. Диастолическое давление** Минимальное 9.45(**ВЫБРОС?**), Максимальное 140(**ВЫБРОС?**) Среднее ≈77.1, Медианное ≈78
И Т.Д.
